Switch from association to link

tmadlener opened the pull request:
BEGINRELEASENOTES
- Switch from `Association` to `Link` in the conversion from EDM4hep to LCIO to accomodate for the new naming from ([EDM4hep#341](https://github.com/key4hep/EDM4hep/pull/341))
- Add test to make sure that conversion of Links from EDM4hep to LCIO works as expected

ENDRELEASENOTES

Fixes #199 (I think). @andresailer was this what you were referring to in that issue?
Fixes https://github.com/key4hep/k4EDM4hep2LcioConv/issues/93

jmcarcell commented on k4MarlinWrapper/src/components/EDM4hep2Lcio.cpp:
```diff
@@ -381,13 +381,13 @@ StatusCode EDM4hep2LcioTool::convertCollections(lcio::LCEventImpl* lcio_event) {
 
   for (const auto& [edm4hepName, lcioName] : collsToConvert) {
     const auto coll = getEDM4hepCollection(edm4hepName);
-    if (coll->getTypeName().find("Association") != std::string_view::npos) {
```
This is what #199 is referring to since now it's impossible to find "Association" in the type name.

tmadlener commented:
Seems, I didn't merge this yet. AFAIU this solves the issue and can be merged, right?

jmcarcell commented:
Maybe there should be a test doing something with links? Otherwise this can just not be caught if there is another problem like this one.

tmadlener commented:
There is already a test with links [`global_converter_maps`](https://github.com/key4hep/k4MarlinWrapper/blob/main/test/README.md#global_converter_maps) that goes from LCIO to EDM4hep (which works without problems apparently). I added a new test that also checks the other direction. I added this as a completely separate test that uses one new `Transformer` for creating the `Links` in EDM4hep and then it uses another new Marlin `Processor` for checking it's contents. Prior to the fixes this test fails, so it does indeed check that links are converted.

tmadlener commented on test/src/MarlinMCRecoLinkChecker.cc:
```diff
@@ -0,0 +1,89 @@
+
+  marlin::Processor* newProcessor() override { return new MarlinMCRecoLinkChecker; }
+
+  /** process the event - In this case simply link MCparticle[i] with
+   * ReconstructedParticle[i]
+   */
```
```suggestion
```

tmadlener commented:
@jmcarcell do you want to have another closer look? Otherwise, I would merge this later today.

jmcarcell commented on test/src/MarlinMCRecoLinkChecker.cc:
```diff
@@ -0,0 +1,86 @@
+                               std::to_string(i) + ", actual: " + std::to_string(rel->getWeight()) + ")");
+    }
+
+    if (!(rel->getTo() == mc)) {
```
So we don't have `!=`?

tmadlener replied:
We do, especially because for LCIO this is just comparing pointers. (We would also have it for EDM4hep). This is a remnant from copy-pasting.
